docs: :memo: split guide into creating vs managing a Data Package (#1592)
seedcase-project/seedcase-sprout: docs/guide/packages.qmd

---
title: "Creating a Data Package"
order: 1
jupyter: python3
---

> ## What is a Data Package

Reviewer: ```suggestion
## What is a Data Package?
```

This sentence feels more like a question.

At the core of Sprout is the [Data Package](/docs/glossary.qmd), which is

> a standardized way of structuring and documenting data. Technically, a data

Reviewer: Hmmmmmmmmmmmmm that's a good point :thinking: I like your suggestion in the last sentence, go with that :grin:

> package is any folder that includes a file called `datapackage.json` containing

Reviewer: ```suggestion
Package is any folder that has a file called `datapackage.json` in it that contains
```

Data Package is capitalised to match how they write it.

metadata following the [Frictionless Data Package
specification](https://datapackage.org/standard/data-package/#introduction).

> The metadata in `datapackage.json` describes both the project (e.g. its title,

Reviewer 2: ```suggestion
The metadata in `datapackage.json` can describe both the Data Package (e.g. its title,
```

creation data, and contributors) and any data included in the project (e.g. the
name, data type, and description of each column/variable in the data). This
metadata is stored in JSON format with a standardized naming convention, e.g.
the title of the project should be stored as `"title": "your project title"`.
In this guide, we will refer to each piece of metadata stored in the JSON file
as a [*metadata property* (or just a *property* for
short)](/docs/glossary.qmd). You can [view an example of a complete
`datapackage.json` file in our example
repository](https://github.com/seedcase-project/example-seed-beetle/blob/main/datapackage.json).

> ## Creating a Data Package with Sprout

Author: Yes, I can see that would make sense as well. It seemed like a bigger makeover so I left it like this. One thing I do like with the current approach is that the initial file structure is smaller so it is easier to overview and understand what everything does which has pedagogical benefits, but I also understand the reasons for wanting people to use the template right away.

> ## Creating a Data Package with Sprout

Reviewer: Good point. I'll make an issue about that :+1: Related to the simpler, that just means we should have a simpler option in the template :grin:

> ## Creating a Data Package with Sprout

Reviewer 2: I generally I'm not the biggest fan of two headers directly after each other. Maybe add a short meta section about the steps that the subsections will cover? :)

### Creating a script to help manage metadata

A standardized naming scheme has many advantages in terms of reproducibility,
but it also makes it difficult to manually create the `datapackage.json` file
since it is difficult to remember all the naming conventions for the different
metadata properties in the JSON file. Manually editing this file is also
ill-advised since it would make you susceptible to introducing errors via typos

> and misunderstandings of the specification. To make this process more

Reviewer: ```suggestion
potential misunderstandings of the specification. To make this process more
```

convenient and robust, Sprout provides a way to create and edit the metadata in
`datapackage.json` programmatically via Python scripts.

Let's start by taking a look at what your project's file structure looks like

> at this point:

Reviewer 2: Aren't we missing a step here? When were these files created?

## What is a Data Package?

At the core of Sprout is the [Data Package](/docs/glossary.qmd), which is
a standardized way of documenting data via structured metadata. Technically, a data
Package is any folder that has a file called `datapackage.json` in it that contains
metadata following the [Frictionless Data Package
specification](https://datapackage.org/standard/data-package/#introduction).

The metadata in `datapackage.json` can describe both the Data Package (e.g. its title,
creation data, and contributors) and any data included in the Data Package (e.g. the
name, data type, and description of each column/variable in the data). This
metadata is stored in JSON format with a standardized naming convention, e.g.
the title of the project should be stored as `"title": "your project title"`.

In this guide, we will refer to each piece of metadata stored in the JSON file
as a [*metadata property*](/docs/glossary.qmd) (or just a *property* for
short). You can view an example of a complete
`datapackage.json` file in this [example

> repositories](https://github.com/seedcase-project/example-seed-beetle/blob/main/datapackage.json).

Reviewer 2: ```suggestion
repository](https://github.com/seedcase-project/example-seed-beetle/blob/main/datapackage.json).
```

## Using Sprout to create a Data Package

The Data Package standard provides a standardised structure and naming scheme in `datapackage.json`.
This has many advantages in terms of reproducibility,
but it also makes it difficult to manually create the `datapackage.json` file
and remember all the names for each of the different properties. Manually editing this file is also
ill-advised since you are more likely to accidentally make typos or accidentally not follow the schema provided by the specification. To make this process more
convenient and robust, Sprout provides a way to create and edit the metadata in
`datapackage.json` programmatically via Python scripts.

### Creating the package properties script to manage your metadata

Let's start by taking a look at what your project's file structure should look
like after following [the installation section of this

> guide](https://sprout.seedcase-project.org/docs/guide/installation):

Reviewer 2: There's something that's off in terms of the order/continuity of the guides with this. At the beginning, we refer to what we did in the previous guide, then we delete some of what we did there, and now we refer to the file structure looking as if the project was just initialised. Can this be improved somehow? :)

```

::: callout-important
Sprout assumes the working directory is the root of your Python
project---where your `.git/` folder and/or `pyproject.toml` file are located.
If you used our Data Package [template](https://template-data-package.seedcase-project.org), you will see additional config files
to the ones listed above.
:::

Since there is no `datapackage.json` file in our root folder, we can't yet call
it a Data Package. So let's embark on the journey of creating this file,

> starting with understanding the content of `main.py`. The `main.py`

Author: Haha, I agree with you and was thinking that it was a bit grandiose as I was writing it as well. I made a minor edit to your suggestion deleting the word "though" and incorporated it.

    main()
```

Then, use uv to run the script from the Terminal:

``` {.bash filename="Terminal"}
uv run main.py
```

<!-- Create the script where quarto can find it for building the docs -->
```{python}
#| include: false
sp.create_properties_script(package_path.root())
```

This will create a `package_properties.py` file in the newly created `scripts/`
folder of your Data Package.

<!--TODO could this callout come later? -->

> <!--TODO If the __init__ file is required in all cases, why is its creation no automated? -->

Author: ```suggestion
```

print(file_tree(package_path.root()))
```

### Managing metadata programmatically

Inside the `scripts/package_properties.py` file, you will find a template for
creating the metadata properties of your Data Package programmatically. We will

> walk through the full template in more detail in the next section of the guide.

Reviewer 2: Maybe add a link to that next "section" 
```suggestion
walk through the full template in more detail in the [next section](./docs/guide/manage-packages.qmd) of the guide.
```

Here, we will first start with a simplified `package_properties.py` script to
easier understand how to manage metadata via this script.

> As you can see in the code snippet below, Sprout uses Python classes to

Reviewer: ```suggestion
Sprout uses Python classes to
```

Actually isn't even necessary to say this, as the code chunk below is relevant to the fact that Sprout uses classes.

> facilitate the entry of metadata. For example, the `PackageProperties` class

Reviewer: ```suggestion
support creating and editing metadata. For example, through the `PackageProperties` class
```

allows you to use tab completion to view the names of all existing metadata
properties instead of trying to memorize them or looking them up in the
documentation each time. The docstring of each class has additional information
of how the metadata should be entered; it is available both in your IDE and

> online, e.g. for [`PackageProperties`](/docs/reference/PackageProperties.qmd).

Reviewer: ```suggestion
this website, e.g. for [`PackageProperties`](/docs/reference/PackageProperties.qmd).
```

<!--TODO Why isn't the class used to checks the types directly? Because it

> involves a reimplementation of the schema? -->

Reviewer: Yes :stuck_out_tongue: that's way outside the scope of this package. And way too much work. We started down that path but quickly decided against it.

Author: ```suggestion
```

Makes sense!

In the example below, we have filled out all the required metadata properties
on the package level with sample values. You can go ahead and copy this over
to your `package_properties.py` replacing any existing code there.

```{python}
#| filename: "scripts/package_properties.py"

import seedcase_sprout as sp

package_properties = sp.PackageProperties(
    name="diabetes-study",
    title="A Study on Diabetes",

>     description="Data from a 2021 study on diabetes prevalence",

Author: I wanted a simpler example first and moved the longer text to later in the guide

:::

The `write_properties()` function created the `datapackage.json` file in
your Data Package's `diabetes-study` folder, which contains the
properties you added to it. Now, you will see the added
`datapackage.json` file in your Data Package folder.

```{python}
#| echo: false
print(file_tree(package_path.root()))
```

The content of this file should look like this:

```{.json filename="datapackage.json"}
{
  "name": "diabetes-study",
  "id": "c0f9b217-589b-4ee4-a917-3c4f21e3be8d",
  "title": "A Study on Diabetes",
  "description": "Data from a 2021 study on diabetes prevalence",
  "version": "0.1.0",
  "created": "2025-11-05T16:16:20+01:00",
  "licenses": [
    {
      "name": "ODC-BY-1.0"
    }
  ]
}

> ```

Reviewer: This can be auto-created using print/bash cat, as it can quickly get out of date with changes above.

Congratulations, you have created your first Data Package! Remember that you

> will never need to edit the `datapackage.json` file manually. Instead, you will

Reviewer: ```suggestion
will never need to edit the `datapackage.json` file manually. Instead, you'll
```
